pyasn1/type/univ.py

```diff
@@ -2057,16 +2075,24 @@ def __init__(self, **kwargs):
         self._dynamicNames = self._componentTypeLen or self.DynamicNames()
 
     def __getitem__(self, idx):
-        if octets.isStringType(idx):
-            return self.getComponentByName(idx)
-        else:
-            return base.AbstractConstructedAsn1Item.__getitem__(self, idx)
+        try:
+            if octets.isStringType(idx):
+                return self.getComponentByName(idx)
+            else:
+                return self.getComponentByPosition(idx)
+
+        except error.PyAsn1Error:
+            raise KeyError(sys.exc_info()[1])
```

Comment:
That's a problem here. Sequence in ASN.1 is considered an [ordered list](https://www.obj-sys.com/asn1tutorial/node11.html), which I think means it should raise `IndexError`.

`collections.Sequence` should definitely raise `IndexError` - that's clear because it has an `index()` function and is reversible, so the order of components is important.

`collections.Mapping` should definitely raise a `KeyError`, because it's not ordered, and any object can be used as a key.

`collections.Set` is more complex - it simply does not support `__getitem__` at all. It just supports `__contains__`.

So I think having a `SequenceAndSetBase` is problematic, because a) `__getitem__` should only apply to Sequence, and b) it should then raise `IndexError` and not `KeyError`.

I hope that makes sense!

Reply:
>  So I think having a SequenceAndSetBase is problematic, because a) __getitem__ should only apply to Sequence, and b) it should then raise IndexError and not KeyError.

Bear in mind that ASN.1 `SEQUENCE` type corresponds to Python `collections.Mapping`, while the ASN.1 `SEQUENCE OF` type resembles Python `collections.Sequence`.

You may agree that the ASN.1 `Sequence` object should therefore raise `KeyError` where ASN.1 `SequenceOf` object should raise `IndexError`.

On top of that in ASN.1 we have another pair of types (a mapping and a sequence) with additional property -- their components are unordered but, unlike Python `set`, the *values* contained in ASN.1 `SET` are not necessarily unique (though the ASN.1 types should be distinct to be distinguishable).

Now, you may agree that ASN.1 `SET`/`SET OF` could as well be modelled by Python `Mapping` and `Sequence` respectively -- this is how pyasn1 is trying to approach it.

Does it make sense?

Reply:
I don't think that's quite right.

If I take e.g. https://www.obj-sys.com/asn1tutorial/node11.html as a simple enough documentation point, then `SEQUENCE OF` requires all items to be of the same type. That's simply not something that Python implements. All `collections.Sequence` can contain any object; they just have different other properties such as reversibility or mutability, etc.

Similarly, `SEQUENCE` is described as an ordered list. That's exactly what Python `list` and `tuple` types are, which are both `collections.Sequence`.

So I'm not agreeing with that first part, unless I misunderstand something fundamentally about ASN.1. Which is possible, but then I'd like to learn and urge you to point me to a source that explains `SEQUENCE` and `SEQUENCE OF` behaviour in the way you describe.

The same page describes `SET` as an *unordered* collection, which has no real equivalence in Python ABCs. `collections.Set` can be serialized as ASN.1 `SET`, but deserializing ASN.1 `SET` into `collections.Set` would drop duplicates, so is probably not the best approach.

`SET OF` appears to be `SET` with a single permissible value type, much like `SEQUENCE OF` is to `SEQUENCE`. That finds even less of an equivalent in Python. You might look to e.g. https://pypi.python.org/pypi/multiset for an equivalent to ASN.1 `SET`.

But again, please, if I grossly misunderstand ASN.1, I'd like to learn! This is my current understanding, however, and by that understanding I can't agree with your implementation choice.

Reply:
As a bit of background, my https://github.com/jfinkhaeuser/bran project is not about implementing ASN.1 perfectly. Instead it is about *using* ASN.1 for providing a canonical serialization of Python types.

I do that by mapping various Python types to ASN.1 types (mostly `SEQUENCE`, really); I can guarantee that there will be exactly one byte representation for a given Python value.

That means I really only need to understand this particular byte representation and deserialize it into a Python type matching the pre-serialization type. The implication, for example, is that it's perfectly safe for me to serialize `sequence.Set` to ASN.1 `SET`, and then drop duplicates during deserialization because the original value could not have contained duplicates.

That makes my view of ASN.1 very Python-centric. Yours is probably very ASN.1 centric, or so I presume.

Reply:
You've raised a very interesting point!

The pyasn1 tool has been hugely influenced by [this book](http://www.oss.com/asn1/resources/books-whitepapers-pubs/dubuisson-asn1-book.PDF). You may find many references to the contents of this book in the code (e.g. design and terminology).

Back to our `Mapping` vs `SEQUENCE` and `Sequence` vs `SEQUENCE OF` discussion. I do agree that my use of "correspondence" is way too stretchy -- these ASN.1 types arguably share **some** behaviour properties with the Python types, but they also differ a great deal.

Here's what the above mentioned book says on ASN.1 `SEQUENCE`:

```
The SEQUENCE constructed type  that is the most commonly used by  computing 
languages is that which provides an ordered series of elements of different types.
...
The SEQUENCE type would be no more than the equivalent of struct type to be
found in numerous languages.
...
```

And `SEQUENCE OF`:

```
The SEQUENCE OF type  is equivalent  to the dynamic arrays or lists of some programming
languages: all the elements are of the same type but the number of the elements is not 
necessarily known beforehand .
...
```

At the same time, `SEQUENCE` behaves like a `list` if you ignore the named fields.

The difference between Python and ASN.1 types that I can see is that, as a language, ASN.1 is typed stronger than Python is. As a practical example, consider Python `set` where you can't have two identical values (e.g. `set([1, 1]) == set([1])`) but you still can have identical types (e.g. `set` with distinct `int` values) whereas in ASN.1 you can't even have two identical types in `SET` what also implies no identical values.

In ASN.1 types are (arguably) identified by the concept of `TAG` and (also arguably) imposed constraints on the allowed values. Type matching in pyasn1 is solely based on tags (`.tagSet`) and value constraints (`.subtypeSpec`), not on Python types.

As you say, many ASN.1 types features do not have reasonable equivalent among the canonical Python types/protocols. So pyasn1 tries to model those ASN.1 quirks and expose the ASN.1-specific type's behaviour through the `.getComponentBy**`, `.setComponentBy*`, `.isSuperTypeOf` and a bunch of other methods/properties. On the flip side, this wordy API is not Pythonic as you have noticed.

Now the question is - how would you model ASN.1 types in a way to have them Pythonic **and** expose ASN.1 functionality they posses? The best I could come up with so far is to have a thin layer over the base ASN.1 functionality which duck-types Python `dict`, `list`, `string`, `bytes`, `int`, `float` etc for as long as I can find at least some correspondence.

For types like `INTEGER` it's easy. However for `SEQUENCE` it's less obvious, what brings us back to this PR. Given the hybrid nature of ASN.1 `SEQUENCE`, pyasn1 duck-types `dict` and partially `list`:

```
>>> seq = Sequence()
>>> seq['name'] = OctetString('John')
>>> seq.keys()
['name']
>>> seq.values()
['John']
>>> seq['name']
'John'
```

That's `dict` so far, but here is where `list` behaviour sneaks in:

```
>>> seq[0]
'John'
```

There more deviations from the canonical `dict` behaviour in part that only registered keys will be accepted.

> The implication, for example, is that it's perfectly safe for me to serialize sequence.Set to 
> ASN.1 SET, and then drop duplicates during deserialization because the original value could
> not have contained duplicates.

I am not sure. Remember, ASN.1 `SET` has unique types, not values. The ASN.1 `SET` would only let you one type per structure. Like if you have:

```
person = {'age': 42, 'children': 2}
# or even simpler
person = set([42, 2])
```

That won't fit into:

```
person ::= SET {
  age  INTEGER,
  children INTEGER
}
```

You could still try to fit that into `SET OF` for as long as you have all items of the same type (which may not hold true for a Python `set`).

> That makes my view of ASN.1 very Python-centric. Yours is probably very ASN.1 centric,
> or so I presume.

This is wonderful! ;-)

Looking at the `bran` (I really like the project!), would it make sense to stick to ASN.1 `SEQUENCE` for encoding Python `Mapping` and `Sequence`, just tag that `SEQUENCE` differently to distinguish between original Python types it holds?

I am not sure you could find a sufficiently clear correspondence between ASN.1 and Python types -- it's not just implementation shortcomings, the type systems were designed with different priorities in mind... 

WDYT?

p.s. One last note: keep in mind that field names of a `SEQUENCE` are never transferred over the wire. If you put `dict` keys in there - they won't pop out at the receiving end! So you may need to pass the keys as a complimentary `SEQUENCE`.